Src/WindowsServer/WindowsServer.Shared/Implementation/EnvironmentVariableMonitor.cs

```diff
@@ -0,0 +1,114 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer.Implementation
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Threading;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+
+    /// <summary>
+    /// Utility to monitor the value of environment variables which may change 
+    /// during the run of an application. Checks the environment variables 
+    /// intermittently.
+    /// </summary>
+    internal abstract class EnvironmentVariableMonitor : IDisposable
+    {
+        // Environment variables tracked by this monitor.
+        protected readonly ConcurrentDictionary<string, string> CheckedValues;
```

Comment:
Why is it a concurrent collection? Do you plan to run multiple instances of timer callback at the same time?

Reply:
For this abstract class I set out to run it on a single timer callback, and I even guard specifically against re-entrancy with how I use the timer. 

However, because it is possible that some subscriber to the event might exist on different threads, I wanted to ensure those cases were covered.

Perhaps I was being over cautious here.

Reply:
@SergeyKanzhelev Would you like me to change this for this PR?